Step 3

ilbender opened the pull request:
i am mostly unsure about the usage (or lack thereof) of subscribeOn and observeOn in FeedRepository's refresh, getAllProjects and FeedViewModel's init and refresh

ilbender committed:
1. changed FeedApiResponseGenerator so that theres no need for a getter.
2.renamed few methods in FeedViewModel
3.renamed columns as per snake case conventions
4. renamed getUserCount in DataAccessObjects.kt to getUserProjectCount
5. changed getById query

ilbender committed:
started building tests

ilbender committed:
started building tests

ilbender committed:
changed viwModel according to the new explanation

ilbender committed:
created tests to test both repo's ability to save data and model actually getting the data from the repo.
can pass mockApi in the constructor.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt:
```diff
@@ -1,9 +1,48 @@
+    fun getAllProjects(): Observable<List<UserProject>> {
+
+        return feedDatabase.userProjectDao().getAll()
+            .subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
```
Unfortunately, in RX there is no rule regarding who (the caller or the callee) should specify `subscribeOn()` and `observeOn()`.

In this case `FeedRepository` knows that reading from a DB should be done on the IO thread, so it's fine to say `subscribeOn(Schedulers.io())`.

On the other hand, the line `observeOn(AndroidSchedulers.mainThread())` assumes that whoever calls `repository.getAllProjects()` has to handle the result on the UI thread. And this assumption in not justified.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt:
```diff
@@ -1,9 +1,48 @@
+
+
+    fun refresh(): Completable {
+        return feedApiService.getFeedData().subscribeOn(Schedulers.io()).flatMapCompletable { feedData ->
```
Long lines with multiple actions are hard to follow. 

Compare this:
`service.extract(a, b, c).transform(d, e, f, g).load(h, i, j)`

with this:
```
service.extract(a, b, c)
       .transform(d, e, f, g)
       .load(h, i, j)
```

Besides they look like a train :)
`Locomotive.TrainWagon.FlatWagon.TankWagon.HopperWagon.SchnabelWagon`

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt:
```diff
@@ -1,9 +1,48 @@
+    }
+
+    private fun rotateList(originalList : List<UserProject>) : List<UserProject>{
+        val currentRotations = this.rotationNumber % originalList.size
```
Bonus point!

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedFragment.kt:
```diff
@@ -51,8 +55,12 @@ class FeedFragment : Fragment() {
+    private fun buildDb(): FeedDatabase {
+        val DB_NAME = "user_project_db.db"
+        val db = Room.databaseBuilder(requireContext().applicationContext as Application, FeedDatabase::class.java, DB_NAME).build()
+        return db
```
This can be inlined.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -77,11 +46,31 @@ open class FeedViewModel(application: Application) : AndroidViewModel(applicatio
+                )
+            }
+        }, { error ->
+            updateState { copy(getFeedItems().value, false) }
```
When using `copy()` you only need to specify the properties that changed.
Did you feed items change because of the error?

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -77,11 +46,31 @@ open class FeedViewModel(application: Application) : AndroidViewModel(applicatio
 
     init {
-        refresh()
+        val disposableA =  feedRepository.getAllProjects().subscribe({ userProjects ->
```
This is another way to do it.

bmax-ltx reviewed (commented):
Improving all the time!

ilbender committed:
1. moved ObserveOn(mainThread) from repo's getAllProjects to model's init.
2. dismantled the train (steam locomotive?!) in repo's refresh
3. inlined the builddb method in the fragment
4. fixed the usage of copy
